espei/error_functions/chemical_potential_optimizer.py

```diff
@@ -0,0 +1,102 @@
+"""
+Author: Jonathan Siegel
+
+Contains classes and functions for optimizing at chemical potentials.
+"""
+
+import numpy as np
+from pycalphad.core.utils import point_sample
+from pycalphad.core.constants import MIN_SITE_FRACTION
+
+class PhaseConditions:
+  """
+  Contains data about a given phase at a given temperature and pressure.
+  Evaluates objectives, gradients and Hessians of the Gibbs Free energy.
+  Also contains information about the sublattices and site fractions.
+  This is passed to the optimization function as an input.
+  """
+  def __init__(self, models, phase, state_vars, obj_f, grad_f, hess_f, parameters):
+    self.state_vars = state_vars
+    self.obj_f = obj_f
+    self.grad_f = grad_f
+    self.hess_f = hess_f
+    self.phase = phase
+    self.parameters = np.array(list(parameters))
+    # Collect information about sublattices and which components they contain.
+    self.site_fraction_counts = []
+    self.site_ratios = {}
+    ind = 0
+    while True:
+      count = len([x for x in models[phase].site_fractions if x.sublattice_index == ind])
+      if count == 0:
+        break
+      self.site_fraction_counts.append(count)
+      ind += 1
+    # Generate matrix mapping site fractions to composition. Remove vacancies.
+    self.composition_conversion_matrix = \
+          np.zeros([len([x for x in models[phase].components if x.number_of_atoms > 0]), \
+                    len(models[phase].site_fractions)])
+    # Collect site ratios.
+    for i, x in enumerate(models[phase].site_fractions):
```

Comment:
Site ratios could also come from the `sublattices` attribute of a `Phase` oject

`dbf.phases[phase_name].sublattices`